internal/errors/cassandra_test.go

```diff
@@ -20,8 +20,688 @@ package errors
+				test.beforeFunc()
+			}
+			if test.afterFunc != nil {
+				defer test.afterFunc()
+			}
+			if test.checkFunc == nil {
+				test.checkFunc = defaultCheckFunc
+			}
+
+			got := NewErrCassandraNotFoundIdentity()
+			if err := test.checkFunc(test.want, got); err != nil {
+				tt.Errorf("error = %v", err)
+			}
+		})
+	}
+}
+
+func TestNewErrCassandraUnavailableIdentity(t *testing.T) {
+	type want struct {
+		want error
+	}
+	type test struct {
+		name       string
+		want       want
+		checkFunc  func(want, error) error
+		beforeFunc func()
+		afterFunc  func()
+	}
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		{
+			name: "returns cassandra unavailable identity error",
+			want: want{
+				want: &ErrCassandraUnavailableIdentity{
+					err: New("cassandra unavailable"),
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(tt *testing.T) {
+			if test.beforeFunc != nil {
+				test.beforeFunc()
+			}
+			if test.afterFunc != nil {
+				defer test.afterFunc()
+			}
+			if test.checkFunc == nil {
+				test.checkFunc = defaultCheckFunc
+			}
+
+			got := NewErrCassandraUnavailableIdentity()
+			if err := test.checkFunc(test.want, got); err != nil {
+				tt.Errorf("error = %v", err)
+			}
+		})
+	}
+}
+
+func TestErrCassandraUnavailable(t *testing.T) {
+	type want struct {
+		want error
+	}
+	type test struct {
+		name       string
+		want       want
+		checkFunc  func(want, error) error
+		beforeFunc func()
+		afterFunc  func()
+	}
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		{
+			name: "returns cassandra unavailable identity error",
+			want: want{
+				want: &ErrCassandraUnavailableIdentity{
+					err: New("cassandra unavailable"),
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(tt *testing.T) {
+			if test.beforeFunc != nil {
+				test.beforeFunc()
+			}
+			if test.afterFunc != nil {
+				defer test.afterFunc()
+			}
+			if test.checkFunc == nil {
+				test.checkFunc = defaultCheckFunc
+			}
+
+			got := ErrCassandraUnavailable()
+			if err := test.checkFunc(test.want, got); err != nil {
+				tt.Errorf("error = %v", err)
+			}
+		})
+	}
+}
+
+func TestErrCassandraNotFound(t *testing.T) {
+	type args struct {
+		keys []string
+	}
+	type want struct {
+		want error
+	}
+	type test struct {
+		name       string
+		args       args
+		want       want
+		checkFunc  func(want, error) error
+		beforeFunc func(args)
+		afterFunc  func(args)
+	}
+	defaultCheckFunc := func(w want, got error) error {
+		if !Is(got, w.want) {
+			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+		}
+		return nil
+	}
+	tests := []test{
+		func() test {
+			keys := []string{
+				"uuid",
+			}
```

Comment:
```suggestion
			keys := "uuid"
```

Reply:
thank you.
I don't think it's necessary to expose the keys, so what about the following?

```
return test{
				name: "returns cassandra key not found error when key is not found",
				args: args{
					keys: []string { 
						"uuid"
					},
				},
				want: want{
					want: New("cassandra key 'uuid' not found: cassandra entry not found"),
				},
			}

```

Reply:
As below 
https://github.com/vdaas/vald/pull/865#discussion_r538934310

Reply:
@hlts2 
but, the function can receive the value of `string` not `[]string`, how do you check it?
This pattern should be tested.
I think if you do not think, you should refactor function.
@kevindiu please confirm it

Reply:
@vankichi 
we cannot declare a variable length variable (like `...string`), so we have to use slice in test case.

Reply:
@kevindiu 
but, a variable length variable get single value, right?

Reply:
@vankichi @kevindiu 
thank you.
I think Variadic arguments are internally represented as an array. So I don't think there is any particular problem.
what do you think?

Reply:
> there is any particular problem
It is not related with this discussion.

I just say we should test all valuable pattern which the function can get.

Reply:
@vankichi but the truth is that, the function can receive multiple values in the same `variable`. 
Also golang treat the variable length as slice. 
For example to use the variable length variable, we loop though this variable `for xxx:= range variable_length_var { ...`, so in my opinion there are no difference using slice in test case.

Reply:
@kevindiu 
I know that.

> I just say we should test all valuable pattern which the function can get.

Reply:
@vankichi 

When testing variadic arguments, there are four argument patterns to test like this.

```go
// Target for test
func fn(s ...int) {
	fmt.Println(s)
}

// There are 4 pattern for test
fn()               // empty value
fn(1)              // single value
fn(1, 1)           // multi values
fn([]int{1, 1}...) // multi values of slice
```

If we want to test it rigorously, we should test it in all cases.
So as you say, I agree with your opinion.
However, covering all test cases can take a lot of time.
And there is also the concern that test cases will become more complex and maintenance costs will increase.
Therefore, there is a method to divide the equivalence(quivalence partitioning method) and omit the test case.
The Go language has a language specification that variadic arguments can be represented by slices.
So I think that it can be omitted by making the test common by using its use.
For example, `fn (1)` can be expressed as `fn ([] int {1} ...)`. and `fn (1, 2)` can be represented by `fn ([] int {1, 2} ...)`. 
And the following is how to omit the Empty pattern.
```go
type args struct {
    num []int
}

a := args {
  num: nil,
}
// or 
a := args {
  num: []int{},
}

// for empty pattern case. 
// `fn ()` can be represented by the `fn(a.num...)  `
fn(a.num...)  
```

If the function to be tested is not complicated like this, I think that it could be omitted by using a slice.
if I made a mistake, I am sorry.

Reply:
In my opinion, rather than testing on `how user use the program`, we should test about how the program behave (business logic). We variable length variable behave like a slice.
Like @hlts2 said, there are so many cases to test, and in my opinion, testing `different methods to pass in the argument` is more like testing the golang itself, so in my opinion, we can skip testing this pattern.

(sorry I couldn't found any official document about the variadic function.)